File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.js
function solution(numbers, hand) {
    let result = ''
    let handwhL = [0, 3];
    let handwhR = [2, 3];
    
    for(let i =0; i < numbers.length; i++)
        {
            if(numbers[i] === 1)
                {
                    result += 'L'
                    handwhL = [0, 0]
                }
            else if(numbers[i] === 4)
                {
                    result += 'L'
                    handwhL = [0, 1]
                }
            else if(numbers[i] === 7)
                {
                    result += 'L'
                    handwhL = [0, 2]
                }
            else if(numbers[i] === 3)
                {
                    result += 'R'
                    handwhR = [2, 0]
                }
            else if(numbers[i] === 6)
                {
                    result += 'R'
                    handwhR = [2, 1]
                }
            else if(numbers[i] === 9)
                {
                    result += 'R'
                    handwhR = [2, 2]
                }
            else if(numbers[i] === 2)
                {
                    let disL = Math.abs(handwhL[1]) + Math.abs(handwhL[0] - 1)
                    let disR = Math.abs(handwhR[1]) + Math.abs(handwhR[0] - 1)
                    if(disL > disR)
                        {
                            result += 'R'
                            handwhR = [1, 0]
                        }
                    else if(disL < disR)
                        {
                            result += 'L'
                            handwhL = [1, 0]
                        }
                    else
                        {
                            if(hand === 'right')
                                {
                                    result += 'R'
                                    handwhR = [1, 0]
                                }
                            else 
                                {
                                    result += 'L'
                                    handwhL = [1, 0]
                                }
                        }
                }
            else if(numbers[i] === 5)
                {
                    let disL = Math.abs(handwhL[1] - 1) + Math.abs(handwhL[0] - 1)
                    let disR = Math.abs(handwhR[1] - 1) + Math.abs(handwhR[0] - 1)
                    if(disL > disR)
                        {
                            result += 'R'
                            handwhR = [1, 1]
                        }
                    else if(disL < disR)
                        {
                            result += 'L'
                            handwhL = [1, 1]
                        }
                    else
                        {
                            if(hand === 'right')
                                {
                                    result += 'R'
                                    handwhR = [1, 1]
                                }
                            else 
                                {
                                    result += 'L'
                                    handwhL = [1, 1]
                                }
                        }
                }
            else if(numbers[i] === 8)
                {
                    
                    let disL = Math.abs(handwhL[1] - 2) + Math.abs(handwhL[0] - 1)
                    let disR = Math.abs(handwhR[1] - 2) + Math.abs(handwhR[0] - 1)
                    if(disL > disR)
                        {
                            result += 'R'
                            handwhR = [1, 2]
                        }
                    else if(disL < disR)
                        {
                            result += 'L'
                            handwhL = [1, 2]
                        }
                    else
                        {
                            if(hand === 'right')
                                {
                                    result += 'R'
                                    handwhR = [1, 2]
                                }
                            else 
                                {
                                    result += 'L'
                                    handwhL = [1, 2]
                                }
                        }
                }
            else if(numbers[i] === 0)
                {
                    let disL = Math.abs(handwhL[1] - 3) + Math.abs(handwhL[0] - 1)
                    let disR = Math.abs(handwhR[1] - 3) + Math.abs(handwhR[0] - 1)
                    if(disL > disR)
                        {
                            result += 'R'
                            handwhR = [1, 3]
                        }
                    else if(disL < disR)
                        {
                            result += 'L'
                            handwhL = [1, 3]
                        }
                    else
                        {
                            if(hand === 'right')
                                {
                                    result += 'R'
                                    handwhR = [1, 3]
                                }
                            else 
                                {
                                    result += 'L'
                                    handwhL = [1, 3]
                                }
                        }
                }
        }
    return result;
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    answer = ''
    arr = [["1", "2", "3"], ["4", "5", "6"], ["7", "8", "9"], ["*", "0", "#"]]
    
    ll = ["1", "4" ,"7"]
    rr = ["3", "6", "9"]
    l_prev = "*"
    r_prev = "#"
    
    for nn in numbers:
        n = str(nn)
        if n in ll:
            l_prev = n
            answer += "L"
            continue
        if n in rr:
            r_prev = n
            answer += "R"
            continue
        level = 0
        r_level = 0
        l_level = 0
        r_dist = 9
        l_dist = 9
        r_index = 0
        l_index = 0
        for i in range(len(arr)):
            if n in arr[i]:
                level = i
            if r_prev in arr[i]:
                r_level = i
                for j in range(len(arr[i])):
                    if arr[i][j] == r_prev:
                        r_index = j
                
            if l_prev in arr[i]:
                l_level = i
                for j in range(len(arr[i])):
                    if arr[i][j] == l_prev:
                        l_index = j
        r_dist = abs(level - r_level) + (1 if r_index != 1 else 0)
        l_dist = abs(level - l_level) + (1 if l_index != 1 else 0)
        print(r_dist, l_dist)
        if r_dist > l_dist:
            l_prev = n
            answer += "L"
        elif r_dist < l_dist:
            r_prev = n
            answer += "R"
        else:
            if hand == "left":
                l_prev = n
                answer += "L"
            else:
                r_prev = n
                answer += "R"
    return answer